
<launch>
   
   <!-- Kinect -->
  <include file="$(find openni_launch)/launch/openni.launch">
  	<arg name="depth_registration" value="True" />
  </include>
  
  <!-- TF FRAMES -->
  <node pkg="tf" type="static_transform_publisher" name="base_to_camera_tf" 
  	args="0.0 0.0 0.0 0.0 0.0 0.0 /base_link /camera_link 100" />
   
   <node pkg="nodelet" type="nodelet" name="standalone_nodelet"  args="manager" output="screen"/>
   
  <!-- sync cloud with odometry and voxelize the point cloud (for fast visualization in rviz) -->
  <node pkg="nodelet" type="nodelet" name="data_odom_sync" args="load rtabmap/data_odom_sync standalone_nodelet">
    <remap from="rgb/image_in"       to="camera/rgb/image_color"/>
    <remap from="depth/image_in"     to="camera/depth_registered/image_raw"/>
    <remap from="rgb/camera_info_in" to="camera/depth_registered/camera_info"/>
    <remap from="odom_in"             to="odom"/>
    
    <remap from="rgb/image_out"       to="data_odom_sync/image"/>
    <remap from="depth/image_out"     to="data_odom_sync/depth"/>
    <remap from="rgb/camera_info_out" to="data_odom_sync/camera_info"/>
    <remap from="odom_out"            to="odom_sync"/>
    
    <param name="queue_size" type="int" value="30"/>
  </node>
  <node pkg="nodelet" type="nodelet" name="points_xyzrgb" args="load rtabmap/point_cloud_xyzrgb standalone_nodelet">
    <remap from="rgb/image"       to="data_odom_sync/image"/>
    <remap from="depth/image"     to="data_odom_sync/depth"/>
    <remap from="rgb/camera_info" to="data_odom_sync/camera_info"/>
    <remap from="cloud"           to="voxel_cloud" />
    
    <param name="queue_size" type="int" value="10"/>
    <param name="voxel_size" type="double" value="0.01"/>
  </node>
  
  <!-- Odometry -->
  <node pkg="rtabmap" type="visual_odometry" name="visual_odometry" output="screen">
  	<remap from="rgb/image"       to="camera/rgb/image_color"/>
  	<remap from="depth/image"     to="camera/depth_registered/image_raw"/>
  	<remap from="rgb/camera_info" to="camera/depth_registered/camera_info"/>
	<remap from="odom" to="odom"/>
    
  	<param name="frame_id" type="string" value="base_link"/>
  	
  	<!-- RTAB-Map's parameters: do "rosrun rtabmap visual_odometry (double-dash)params" to see the list of available parameters. -->
  	<param name="Odom/Type" type="string" value="0"/>            <!-- 0=SURF/SIFT 1=FAST/BRIEF -->
  	<param name="Kp/DetectorStrategy" type="string" value="1"/>  <!-- 0=SURF 1=SIFT, used only when Odom/Type=0 -->
  	<param name="Odom/ResetCountdown" type="string" value="0"/>  <!-- 0=disabled, otherwise countdown before auto-resetting odometry when lost. -->
  </node>
  
  <node pkg="rviz" type="rviz" name="rviz" output="screen"/>
</launch>
